### ISSUE TITLE
Building new column names in pivot_wider()

### ISSUE THREAD

**Issue by mikmart** [tidyverse/tidyr#593]
#586 introduced multiple `values_from` columns in `pivot_wide()` by appending the value column names to the new column names in the output.

It would be nice to have a more general interface to control how the new column names are built: there are many examples of cases where the value names are prepended rather than appended to the column names (e.g. all [examples in the vignette](https://github.com/tidyverse/tidyr/blob/3268807babf9a545cad1560003cab0160d939004/vignettes/pivot.Rmd#L144-L200)), but both methods can be useful.

One possible solution could be to mirror the use of a `.value` sentinel introduced to `pivot_longer()`'s `names_to` in `names_from` for a neat additional symmetry between the pivoting functions:

```r
# for names like: income_estimate, income_moe, ...
us_rent_income %>% 
  pivot_wider(
    values_from = c(estimate, moe),
    names_from = c(variable, .value)
  )

# for names like: estimate_income, estimate_rent, ...
us_rent_income %>% 
  pivot_wider(
    values_from = c(estimate, moe),
    names_from = c(.value, variable)
  )
```

**Comment by mikmart** [comment]
Further, if there was a `names_pattern` (#594), having something like a `names_glue` for constructing the new names with a template passed to `glue()` would again be a nice symmetry:

```r
who_long <- who %>% 
  pivot_longer(
    cols = new_sp_m014:newrel_f65,
    names_to = c("diagnosis", "gender", "age"),
    names_pattern = "new_?(.*)_(.)(.*)",
    values_to = "count"
  )

# to (almost) transform back:
who_long %>%
  pivot_wider(
    names_glue = "new_{diagnosis}_{gender}{age}",
    values_from = "count"
  )
```

**Comment by hadley** [comment]
`names_glue` feels overly flexible to me — I don't think we generally want to encourage people to create column names that can't be parsed with a simple specification (and if you do need to for compatibility with an external tool you can always create the spec by hand).

The biggest challenge with allowing `.value` in `names_from` is not the position of the column, but the difference order of the rows:

```R
us_rent_income %>% 
  pivot_wider_spec(
    values_from = c(estimate, moe),
    names_from = c(.value, variable)
  )
#> 
#>   .name           .value   variable
#>   <chr>           <chr>    <chr>   
#> 1 estimate_income estimate income  
#> 2 estimate_rent   estimate rent    
#> 3 moe_income      moe      income  
#> 4 moe_rent        moe      rent    

us_rent_income %>% 
  pivot_wider_spec(
    values_from = c(estimate, moe),
    names_from = c(variable, .value)
  )
#> 
#>   .name            variable .value  
#>   <chr>            <chr>    <chr>   
#> 1 income_estimate  income   estimate
#> 2 income_moe       income   moe     
#> 3 rent_estimate    rent     estimate
#> 4 rent_moe         rent     moe     
```

And what would it look like if you have `names_from = c("x", ".value", "y")`? — I think this is complex because `values` has to `complete()` all possible combinations in order to produce a rectangle.

And then what if you don't want it ordered that way? I just feels like teaching people to use `complete()` to make their own specification would be simpler.

**Comment by mikmart** [comment]
Yeah I came to realize the order problem as well when I tried to make a start on this.

Naturally a line will have to be drawn somehwere with regard to what can be done outside manual specs, but I think it would be very useful to be able to easily control the value name placement.

I'd say prepended value names come up when reshaping repeated measurements across columns, while appended value names seem to be needed for the situation where the values are statistics that describe a group -- the former is useful for condensing data and may be required for certain analyses, while the latter comes up all the time when reporting data.

I don't have a proposal as to how to solve this right now, but I'll keep thinking about it.
